nomad/state/state_store.go

```diff
@@ -2428,18 +2428,10 @@ func (s *StateStore) UpsertCSIVolume(index uint64, volumes []*structs.CSIVolume)
 				old.Provider != v.Provider {
 				return fmt.Errorf("volume identity cannot be updated: %s", v.ID)
 			}
-
-			// Update fields that are safe to change while volume is being used.
-			if err := old.UpdateSafeFields(v); err != nil {
-				return fmt.Errorf("unable to update in-use volume: %w", err)
-			}
-			v = old
-			v.ModifyIndex = index
-
```

Comment:
It looks like we've moved the logic we wanted here into `Merge`, but then don't we need to add the call to `Merge` here?

Reply:
Nah, we validate and build up new `CSIVolume`(s) in the register/create flows via [`reconcileVolume` -> `Merge`](https://github.com/hashicorp/nomad/blob/4895d708/nomad/csi_endpoint.go#L388-L392) prior to `UpsertCSIVolume` here.  Validation should always happen before getting this far, so "store this in state please" (this upsert method) only handles writing state (and a couple remaining validations above that I opted not to touch).

Reply:
It's that `v = old` line on 2436 that made any updating necessary here (via `UpdateSafeFields`), but that was the source of other intended updates not getting saved.  Without `v = old`, `old` is used only for some last-minute sanity-checking, then the already-merged-earlier `v`olume provided by the caller is what gets committed to state.